### ISSUE TITLE
prop.js error with arrayProp - Type is not a constructor

### ISSUE THREAD

**Issue by spyter** [szokodiakos/typegoose#143]
I am having issues using the @arrayProp() decorator. I have tried it using items and itemsRef and both situations seem to not work for me. If I comment out the property on the model it will compile/run. 

The error that I am getting is in prop.js (line 63: const instance = new Type()). The error is "TypeError: Type is not a constructor". 

Below is an example of what I am attempting to do:

```
export class Test extends Typegoose {
    @prop()
    public name?: string;
}

export class Test2 extends Typegoose {
    @prop()
    public name?: string;

    @arrayProp({ items: Test })
    public tests?: Test[];
}
```

**mongoose version**: _5.0.11_
**typegoose version**: _5.1.0_
**reflect-metadata version**: _0.1.12_
**typescript version**: _2.8.3_
**tsconfig.json** is set to **es2018** _(I tried es5, es6, es2017 too)_

**On a side note**: I am using this with webpack for a node backend server. My app works fine if I comment out the @arrayProp property. 

Has anyone seen this issue before or know how I can get around it? I tried using different versions of Mongoose, but I haven't had any luck. Any suggestions would be much appreciated!

**Comment by marcelh-gh** [comment]
your code example should definitely work. maybe something went wrong during processing with webpack.
there is an issue regarding the arrayProp decorator when used for required properties of non-primitive types though.
[see issue #116](https://github.com/szokodiakos/typegoose/issues/116)

**Comment by spyter** [comment]
Thanks @marcelh-gh for the help, it's very much appreciated. I saw that issue earlier, and I too think they might be related. I tried removing the required constraints from my arrayProp, but it didn't seem to matter. It's possible that it's something related to arrayProp/webpack. I only have the issue when I use the arrayProp attribute. Unfortunately, I need that attribute for my project haha. As of now I haven't found a work around.

**Comment by marcelh-gh** [comment]
@spyter maybe this helps to figure out if you have the same issue..
```
    @prop({ required: true }) // no problem, since no array
    lastname: string;

    @prop({ required: true }) // no problem, since no array
    subdoc: tgSubDoc;

    @arrayProp({ items: String, required: true }) // no problem, since type is primitive
    arrayOfStrings: string[];

    @arrayProp({ items: tgSubDoc }) // no problem, because not declared as required
    optionalArrayOfSubDocs?: tgSubDoc[];

    // *******************************************************************************************************
    @arrayProp({ items: tgSubDoc, required: true }) // PROBLEM, because declared as REQUIRED and ARRAYPROP
    arrayOfSubDocs: tgSubDoc[];
    // *******************************************************************************************************

```

**Comment by spyter** [comment]
@marcelh-gh Thanks again for the help! The first three work for me (even arrayProp with primitive types). But the last two do not. Any array type I have that is not a primitive type, whether it's required or not, does not work. If I comment out that property on my model, it will compile/run without any issues. 

This line, from what I can tell, appears to be where the issue is coming from (or at least that is where it's presenting itself):

**typegoose/src/prop.ts**

**line**: _121_

`const instance = new Type();
`
^^it fails here with "_TypeError: Type is not a constructor_"

Thanks again for the suggestions/help!

**Comment by spyter** [comment]
Ok, so more information on the issue I am having (hopefully it's helpful).

**typegoose/src/prop.ts** - line _191_ 

During runtime, the variable "Type" here is undefined. When **baseProp()** is called, that is why the `const instance = new Type();` is failing for me.

Here is how I am passing it in on my model (_Review_ is just another typegoose model being imported from another file):

```
@arrayProp({ items: Review })
public reviews?: Review[];
```

If I pass in the following, the variable "Type" is no longer undefined on line _191_; however, it's a string (which obviously doesn't work later on). But, it is being passed through as a string. It appears to not work when I am using the class type (above).

```
@arrayProp({ items: "Review" })
public reviews?: Review[];
```

I am using the "ref" (`@prop({ ref: ClassType })`) attribute in places in my models, which appears to be syntactically similar and it is working. 

Any ideas? Thanks :)

**Comment by spyter** [comment]
I found a temporary solution for this. 

The issue appears to be caused by using typescript barrels. If i remove all barrel imports, everything works and the _items_ value for the `@arrayProp()` will work. 

Importing models from other files using their explicit path works (example: `import { Model1} from "./model1.model";`). 

Importing them using barrels (example: `import { Model1, Model2 } from "./";`) does not seem to work.

**Comment by spyter** [comment]
I am still having the issue with enums.

@marcelh-gh does this syntax look correct? 

I am trying to have an array of enum values. I am using enums in other places, but this is the first time I needed to use it as an array. I tried experimenting around with the property options but it didn't seem to help. I am getting the same issue here that I was getting above. This may be a separate issue of its own. If it is I can move it out to another issue but since the underlying problem seems to be similar I thought I'd check here first.

```
enum Permission { One, Two, Three }

@arrayProp({ items: Permission, enum: Permission })
public permissions?: Permission[];
```

Thank you!

**Comment by marcelh-gh** [comment]
@spyter Thanks for sharing your findings. :-) Your issues seem to be different from #116 though. The imports have no effect on that one.
Regarding enums: I am using lots of them without issues. not within arrays though....

**Comment by spyter** [comment]
@marcelh-gh I agree, it's definitely not related to #116 (I thought it was initially) . I am using enums successfully, but I cannot use them in arrays. I'll experiment further with it. Thanks again for your help!

**Comment by Christilut** [comment]
Has anyone found a proper solution for this? This issue comes up a lot in my code and the only "fix" is to re-order imports to prevent a cyclic dependency.
I'd love a better, more permanent, solution to this problem..

**Comment by FSM1** [comment]
Also getting this issue occurring quite often unfortunately and no solution in sight.

**Comment by jfrader** [comment]
No one could figure it out? :(

**Comment by Yackbz** [comment]
I have the same issue using the following code.
How do you reorder the barrels to avoid getting this error?

```javascript
import { IdeaModel } from "../../idea/model/idea.model";
import { arrayProp, instanceMethod, pre, prop, Ref, staticMethod, Typegoose } from "typegoose";
import * as bcrypt from "bcryptjs";
import * as jwt from "jsonwebtoken";

@pre<UserModel>("save", async function(next) {
	// run function on UserModel.save()
	this.password = await bcrypt.hash(this.password, 10);
	next();
})
export class UserModel extends Typegoose {
	@prop({ unique: true })
	username: string;

	@prop()
	password: string;

	@arrayProp({ itemsRef: IdeaModel })
	ideas?: Ref<IdeaModel>[];

	@prop({ ref: IdeaModel })
	bookmarks: Ref<IdeaModel>[];
...
```

**Comment by jaym910** [comment]
I'm getting the same exact issue as well. Any fixes??

**Comment by HorizonShadow** [comment]
Using classes for references is a common problem with ORMs. The solution is to use a string representation of the model.

Just modifying @Yackbz's post...

```
import { IdeaModel } from "../../idea/model/idea.model";
import { arrayProp, instanceMethod, pre, prop, Ref, staticMethod, Typegoose } from "typegoose";
import * as bcrypt from "bcryptjs";
import * as jwt from "jsonwebtoken";

@pre<UserModel>("save", async function(next) {
	// run function on UserModel.save()
	this.password = await bcrypt.hash(this.password, 10);
	next();
})
export class UserModel extends Typegoose {
	@prop({ unique: true })
	username: string;

	@prop()
	password: string;

	@arrayProp({ itemsRef: { name: "IdeaModel" }})
	ideas?: Ref<IdeaModel>[];

	@prop({ ref: { name; "IdeaModel" }})
	bookmarks: Ref<IdeaModel>[];
...
```

**Comment by gentleShark** [comment]
@HorizonShadow thanks for the suggestion but passing itemsRef an a object with a name property assigned the Typegoose class as a string didn't work. 
@spyter the problem persists in my case regardless of barrel imports or directly from source file

**Comment by HorizonShadow** [comment]
@gentleShark Right. The problem is that you're using a class as a reference before it exists.

If A has a reference to B, and B has a reference to A, then either A or B isn't going to exist when creating the class. 

I'm referencing by name as the default for all my reference and it works like a treat.

**Comment by hasezoey** [comment]
@Ben305 this can be closed as "solved"

added documentation in https://typegoose.github.io/typegoose/guides/advanced/reference-other-classes/

**Comment by gentleShark** [comment]
> @Ben305 this can be closed as "solved"
> 
> added documentation in https://hasezoey.github.io/typegoose/guides/advanced/reference-other-classes/

@hasezoey fyi, got a 404 when I followed the link

**Comment by hasezoey** [comment]
> @hasezoey fyi, got a 404 when I followed the link

because the github repo changed (to typegoose), the link is now https://typegoose.github.io/typegoose/guides/advanced/reference-other-classes/ , sorry

**Comment by gentleShark** [comment]
> > @hasezoey fyi, got a 404 when I followed the link
> 
> because the github repo changed (to typegoose), the link is now https://typegoose.github.io/typegoose/guides/advanced/reference-other-classes/ , sorry

good to know! no worries

**Comment by henriquetatagiba** [comment]
@hasezoey what about subdocuments like `@arrayProp({ items: Test })`?

**Comment by hasezoey** [comment]
@henriquetatagiba the same should be working for arrayProp (put it as a string instead of a reference)
-> if it dosnt work, please open a new issue

**Comment by hasezoey** [comment]
@Ben305 this can be closed as "stale" & "fixed?"

if anyone still has this problem in version 6.0.0 or higher, please open a new issue at the new repo

**Comment by Laubeee** [comment]
> Using classes for references is a common problem with ORMs. The solution is to use a string representation of the model.
> 
> Just modifying @Yackbz's post...
> 
> ```
> ...
> 	@arrayProp({ itemsRef: { name: "IdeaModel" }})
> 	ideas?: Ref<IdeaModel>[];
> 
> 	@prop({ ref: { name: "IdeaModel" }})
> 	bookmarks: Ref<IdeaModel>[];
> ...
> ```

I used this workaround and it introduced some weird behaviour in a totally different class where I still used `@prop({ ref: IdeaModel })`
It resulted in not being able to filter for this reference, e.g. `SomeModel.find({ ideaModel: "<ObjID>" })` didnt work. it worked when I used names also in that other class.

Bottom line: be careful with this workaround! probably its best to replace it **everywhere**, even if you don't get any more errors

@hasezoey ping, just in case you want to add this to the docs you referenced above.
